Review

IM-GYURI opened the pull request:
### 변경사항
**AS-IS**
- Seller와 Customer의 회원가입/로그인 기능까지 구현
- common 모듈에 통합되어 있던 ErrorCode와 CustomException

**TO-BE**
- ErrorCode와 CustomException을 interface로 수정 후 common, customer-api, seller-api 모듈 하에 각각 구현하도록 함
- Customer 
  - 고객 정보 조회, 고객  정보 수정, 고객  탈퇴 기능 추가
    - 해당 Customer 본인만 가능하도록 함
  - 이름 제한 : 2 ~ 10자
  - 전화번호 포맷팅
  - @NotBlank 어노테이션에 message 추가
- Seller
  - 판매자 정보 조회, 판매자 정보 수정, 판매자 탈퇴 기능 추가
    - 해당 Seller 본인만 가능하도록 함
  - 이름 제한 : 2 ~ 10자
  - 전화번호 포맷팅
  - @NotBlank 어노테이션에 message 추가
- Product
  - 상품 엔티티 작성
  - 상품 등록, 상품 수정, 상품 삭제 기능 추가
    - 해당 Seller 본인만 가능하도록 함
    
**궁금한 점**
- GlobalExceptionHandler를 common 모듈 하에 작성하고, customer-api와 seller-api가 common을 scan하도록 BeanConfig에 설정해두었음에도 불구하고 SignatureException에 대한 핸들링이 제대로 되고 있지 않는 것 같습니다. 헤더의 Authorization에 잘못된 토큰이 들어올 경우 처리를 하고 싶은데 방법이 있을까요?
- 해당 Customer/Seller 본인만 조회, 수정 등이 가능하도록 Service에 구현해두었는데 제대로 구현한 것인지 모르겠습니다. 제가 작성한 코드보다 더 좋은 방법이 있을까요?
- 어떤 코드가 Service에 작성되어야 하고, Controller에 작성되어야 하는지 구분짓는 것이 조금 헷갈립니다. Controller의 코드가 길어지면 불안한데, 작성한 코드에서 Service로 옮겨야 할 것들이 있을까요?

IM-GYURI committed:
Merge pull request #1 from IM-GYURI/review

docs : update readme.md

IM-GYURI committed:
Merge pull request #2 from IM-GYURI/review

IM-GYURI committed:
feat : add Pattern of name and phone number

name Pattern = length of 2 ~ 10
phone number Patter = 010-0000-0000

IM-GYURI committed:
refactor : delete space

IM-GYURI committed:
feat : add Pattern of name and phone number

name Pattern = length of 2 ~ 10
phone number Pattern = 010-0000-0000

add message for address

IM-GYURI committed:
refactor : delete space

IM-GYURI committed:
feat : get information myself

IM-GYURI committed:
feat : add properties

phone, address, point, birth, createdAt, modifiedAt

IM-GYURI committed:
feat : add findByEmail

find by email and return CustomerDto

IM-GYURI committed:
feat : add SignatureException Handling

IM-GYURI committed:
refactor : change requestMatchers

from new AntPathRequestMatcher to just path string

IM-GYURI committed:
feat : add patch customer information

IM-GYURI committed:
feat : add update customer

IM-GYURI committed:
feat : add find by customer Key

IM-GYURI committed:
feat : add edit customer information

IM-GYURI committed:
docs : check done things

IM-GYURI committed:
feat : add Dto for Edit

IM-GYURI committed:
feat : add delete customer

IM-GYURI committed:
feat : add delete customer and validation by customerKey

IM-GYURI committed:
feat : add delete by customer key and exists by customer key

IM-GYURI committed:
feat : add get information, edit information, delete information of seller

IM-GYURI committed:
feat : add get information, edit information, delete information of seller

IM-GYURI committed:
feat : add properties

IM-GYURI committed:
feat : add updateSeller

IM-GYURI committed:
feat : add existsBySellerKey, findBySellerKey, deleteBySellerKey

IM-GYURI committed:
feat : add findByEmail, edit, delete, validateSellerExistsBySellerKey

IM-GYURI committed:
feat : add Dto for Edit

IM-GYURI committed:
feat : add Category enum

book, clothes, cosmetic, digital, food, furniture, household goods, kitchen utensils, pet goods, sports

IM-GYURI committed:
feat : add Product Entity

id, product key, seller key (FK), name, category, price, stock, description

IM-GYURI committed:
refactor : move check authentication

from controller to service

IM-GYURI committed:
refactor : move check authentication

from controller to service

IM-GYURI committed:
docs : check done things

IM-GYURI committed:
feat : add PreAuthorize

IM-GYURI committed:
refactor : delete AntPathRequestMatcher

IM-GYURI committed:
feat : add builder and setSeller

build ProductEntity
set SellerEntity

IM-GYURI committed:
docs : check done things

IM-GYURI committed:
feat : add findBySellerKey

IM-GYURI committed:
feat : add common EnumValidator

IM-GYURI committed:
feat : add common ValidEnum

IM-GYURI committed:
feat : add product registration

IM-GYURI committed:
feat : add Dto for Product

IM-GYURI committed:
feat : add Product Repository

IM-GYURI committed:
feat : add product registration

IM-GYURI committed:
feat : add Dto for registration

IM-GYURI committed:
feat : add Error Codes for Product

IM-GYURI committed:
feat : add edit product information

IM-GYURI committed:
feat : add update product

IM-GYURI committed:
feat : add find by productKey

IM-GYURI committed:
feat : add edit, findByProductKeyOrThrow, validateAuthorizationAndGetSeller

IM-GYURI committed:
feat : add @Valid

IM-GYURI committed:
feat : add Dto for Edit Product

IM-GYURI committed:
docs : check done things

IM-GYURI committed:
feat : add delete product

IM-GYURI committed:
feat : add deleteByProductKey

IM-GYURI committed:
feat : add delete product

IM-GYURI committed:
feat : change error message

from SELLER NOT FOUND to IS NOT SAME SELLER

IM-GYURI committed:
refactor : change CustomException to CommonCustomException

IM-GYURI committed:
refactor : change CustomException to CustomerCustomException

IM-GYURI committed:
style : delete blank

IM-GYURI committed:
refactor : change ErrorCode to CustomerErrorCode

IM-GYURI committed:
refactor : change from class to interface

IM-GYURI committed:
refactor : change ErrorCode to CustomerErrorCode

IM-GYURI committed:
refactor : change ErrorCode to SellerErrorCode

IM-GYURI committed:
refactor : change enum to interface

IM-GYURI committed:
refactor : change CustomException to CommonCustomException

IM-GYURI committed:
refactor : change CustomException to SellerCustomException

IM-GYURI committed:
refactor : change ErrorCode to SellerErrorCode

IM-GYURI committed:
refactor : change ErrorCode to CommonErrorCode

IM-GYURI committed:
refactor : change CustomException to SellerCustomException

IM-GYURI committed:
refactor : change ErrorCode to SellerErrorCode

IM-GYURI committed:
refactor : CommonCustomException implements CustomException

IM-GYURI committed:
refactor : CustomerCustomException implements CustomException

IM-GYURI committed:
refactor : SellerCustomException implements CustomException

IM-GYURI committed:
refactor : CommonErrorCode implements ErrorCode

IM-GYURI committed:
refactor : CustomerErrorCode implements ErrorCode

IM-GYURI committed:
refactor : SellerErrorCode implements ErrorCode

IM-GYURI committed:
Merge pull request #4 from IM-GYURI/review

IM-GYURI committed:
refactor : move error codes about product

IM-GYURI committed:
feat : add /product/{productKey}

IM-GYURI committed:
refactor : move error codes about product

IM-GYURI committed:
refactor : rename as SellerProductController

IM-GYURI committed:
refactor : rename as SellerProductService

spring-kotlin-lover commented on customer-api/src/main/java/zerobase/customerapi/controller/CustomerController.java:
```diff
@@ -46,4 +54,71 @@ public ResponseEntity<?> signIn(@RequestBody @Valid CustomerSignInDto signInDto)
+      CustomerDto customerDto = customerService.validateAuthorizationAndGetSeller(customerKey,
+          token);
+      return ResponseEntity.ok(customerDto);
+    } catch (CustomerCustomException e) {
+      return ResponseEntity.status(403).body("IS NOT SAME CUSTOMER");
+    }
```
controller 에 있는 catch 구문들은 모두 ExceptionHandler 를 통해 처리되도록 하는 것이 좋을 것 같아요. 대부분 로직들이 중복되고 있고, 예외단위별로 error response 가 정해지기 때문입니다.


 common 모듈에 있는 Handler 를 scan 하도록 하거나, 각 모듈별로 ExceptionHandler 를 별도로 만드는 방법등이 있을 것 같습니다.

spring-kotlin-lover commented on common/src/main/java/zerobase/common/exception/CommonCustomException.java:
```diff
@@ -0,0 +1,22 @@
+import lombok.Getter;
+
+@Getter
+public class CommonCustomException extends RuntimeException implements CustomException {
```
CustomException 을 interface 로 굳이 정의해주신 이유가 있을까요? 전체적인 상속관계를 보았을 때는 CommonCustomException 만 유지되면 괜찮아 보이긴 합니다.

spring-kotlin-lover commented on seller-api/src/main/java/zerobase/sellerapi/service/SellerProductService.java:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  public ProductDto validateAuthorizationAndGetSeller(String productKey, String token) {
+    if (token != null && token.startsWith("Bearer ")) {
+      token = token.substring(7);
+    }
+
+    Authentication authentication = tokenProvider.getAuthentication(token);
+    String email = authentication.getName();
```
token 이 어떻게 넘어오는지는 Http 의 규약인 것으로 보입니다. 따라서, token 정보를 기반으로 authentication 을 가져오는 것은 controller 에서 처리하도록 수정하거나, 혹은 argument resolver 기반으로 처리해보는 방향은 어떨까요?

spring-kotlin-lover commented:
> GlobalExceptionHandler를 common 모듈 하에 작성하고, customer-api와 seller-api가 common을 scan하도록 BeanConfig에 설정해두었음에도 불구하고 SignatureException에 대한 핸들링이 제대로 되고 있지 않는 것 같습니다. 헤더의 Authorization에 잘못된 토큰이 들어올 경우 처리를 하고 싶은데 방법이 있을까요?

GlobalExceptionHandler 를 bean 으로 읽어들이는 방법이 제대로 적용이 안된 것 같기도 합니다. 각 API 단위로 GlobalExceptionHandler class 를 각자 만들어주시거나, bean 을 로드하는 설정에 이상이 없는지 체크해보면 좋겠네요.

> 해당 Customer/Seller 본인만 조회, 수정 등이 가능하도록 Service에 구현해두었는데 제대로 구현한 것인지 모르겠습니다. 제가 작성한 코드보다 더 좋은 방법이 있을까요?

기본 로직은 잘 작성해주셨습니다~! 어떤 부분에서 더 의문점을 갖고 계신가요 ? :) 

> 어떤 코드가 Service에 작성되어야 하고, Controller에 작성되어야 하는지 구분짓는 것이 조금 헷갈립니다. Controller의 코드가 길어지면 불안한데, 작성한 코드에서 Service로 옮겨야 할 것들이 있을까요?

Controller 의 역할과 Service 의 역할이 무엇인지 정리해보면 좋겠습니다. 각 계층별 역할을 명확하게 이해한다면, 어느 로직이 어느 layer 에 위치하는지 알 수 있을 것 같아요. 3 tier layer architecture 의 기본적인 원리와 각 layer 별 역할을 공부해보시면 좋겠네요.

IM-GYURI committed:
feat : delete spring-mvc-pathmatch

IM-GYURI committed:
feat : add ComponentScan

about common and customerapi

IM-GYURI committed:
refactor : delete try - catch

IM-GYURI committed:
feat : delete /product/{productKey}

IM-GYURI committed:
docs : add Trouble Shooting

about @ComponentScan

IM-GYURI committed:
feat : delete spring-mvc-pathmatch

IM-GYURI committed:
refactor : delete BeanConfig

IM-GYURI committed:
refactor : delete try-catch

IM-GYURI committed:
refactor : delete try-catch

IM-GYURI committed:
feat : add @ComponentScan

about common and seller api

IM-GYURI committed:
refactor : delete CustomException

IM-GYURI committed:
refactor : move token processing

from service to controller

IM-GYURI committed:
refactor : delete CustomerCustomException

replace to CommonCustomException

IM-GYURI committed:
refactor : delete CustomerCustomException & move token processing

replace to CommonCustomException &
move token processing from service to controller

IM-GYURI committed:
refactor : delete CustomException

IM-GYURI committed:
refactor : delete ErrorCode

IM-GYURI committed:
refactor : add ErrorCode interface

IM-GYURI committed:
refactor : move token processing

from service to controller

IM-GYURI committed:
refactor : delete SellerCustomException & move token processing

replace to CommonCustomException &
move toke processing from service to controller

IM-GYURI committed:
refactor : delete SellerCustomException

IM-GYURI committed:
refactor : delete try-catch

IM-GYURI committed:
refactor : delete SellerCustomException

replace to CommonCustomException

IM-GYURI committed:
feat : change ddl-auto to update

from create to update

IM-GYURI committed:
refactor : delete BaseEntity

IM-GYURI committed:
refactor : change name from AuthenticationFilter to CustomerAuthenticationFilter

IM-GYURI committed:
refactor : change name of TokenProvider and EditDto by adding 'Customer'

IM-GYURI committed:
refactor : change name by adding 'Customer'

IM-GYURI committed:
refactor : change name by adding 'Customer'

IM-GYURI committed:
refactor : change name by adding 'Customer' and move BaseEntity

IM-GYURI committed:
refactor : change name by adding 'Customer'

IM-GYURI committed:
refactor : change name by adding 'Customer'

IM-GYURI committed:
refactor : change name by adding 'Customer'

IM-GYURI committed:
refactor : move classes related to 'Product'

IM-GYURI committed:
refactor : move BaseEntity

from seller-api to common

IM-GYURI committed:
refactor : move classes related to 'Product'

from seller-api to common

IM-GYURI committed:
refactor : move classes related to 'Product'

from seller-api to common

IM-GYURI committed:
refactor : move BaseEntity to common

IM-GYURI committed:
refactor : move Category to common

IM-GYURI committed:
refactor : move classes related to 'Product' to common

IM-GYURI committed:
feat : add Customer Product Controller

feature of Getting detail information

IM-GYURI committed:
feat : add Customer Product Service

feature of Getting detail information

IM-GYURI committed:
refactor : change name by adding 'Seller'

IM-GYURI committed:
feat : add get product list sorted by name

IM-GYURI committed:
refactor : change name by adding 'Product'

IM-GYURI committed:
refactor : change name by adding 'Product'

IM-GYURI committed:
refactor : change name by adding 'Product'

IM-GYURI committed:
feat : add Dto for sorted Product list

IM-GYURI committed:
feat : add sortedByPriceAsc & Desc, searchByKeyword

IM-GYURI committed:
feat : add findAllByNameContains

IM-GYURI committed:
feat : add searchByKeyword sorted by name, price asc & desc

IM-GYURI committed:
docs : write Trouble Shooting

IM-GYURI committed:
feat : add search by category

sorted by name(asc), price(asc/desc)

IM-GYURI committed:
feat : add search by category

sorted by name(asc), price(asc/desc)

IM-GYURI committed:
feat : add findAllByCategory

IM-GYURI committed:
docs : check done things

IM-GYURI committed:
feat : add recharge point

IM-GYURI committed:
feat : add update point

IM-GYURI committed:
feat : add recharge point

IM-GYURI committed:
feat : add dto for Point recharge

IM-GYURI committed:
feat : add Cart Errors

IM-GYURI committed:
feat : add cart register

cart is automatically created when customer sign up

IM-GYURI committed:
feat : add get cart Information

IM-GYURI committed:
feat : dto for cart

IM-GYURI committed:
feat : add cartEntity

IM-GYURI committed:
feat : dto for cart item

IM-GYURI committed:
feat : add CartItemEntity

IM-GYURI committed:
feat : dto for registering cart

IM-GYURI committed:
feat : add cart repository

existsByCustomerKey, findByCustomerKey

IM-GYURI committed:
feat : add cart register

validateCartExistsByCustomerKey
